### PR TITLE
[FLINK-16373] Make JobManagerLeaderListener thread safe

### PR DESCRIPTION
**Pull request by tillrohrmann**
## What is the purpose of the change

The JobManagerLeaderListener used by the JobLeaderService was not thread safe. Stopping the listener
while notifying a new leader could lead to an IllegalStateException where the rpcConnection which
was supposed to be started was concurrently closed by the stop call.

This PR solves this problem by introducing a lock under which the `JobManagerLeaderListener` calls are executed.

cc @zentol 

## Verifying this change

* Added `JobLeaderServiceTest.handlesConcurrentJobAdditionsAndLeaderChanges`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java [hunk 1]
```diff
@@ -251,90 +254,83 @@ private JobManagerLeaderListener(JobID jobId) {
...
-						} else {
-							LOG.debug("Could not reconnect to the JobMaster {}.", currentRpcConnection.getTargetAddress());
-						}
+			synchronized (lock) {
+				if (stopped) {
+					LOG.debug("Cannot reconnect because the JobManagerLeaderListener has already been stopped.");
+				} else {
+					if (rpcConnection != null) {
+						Preconditions.checkState(
+							rpcConnection.tryReconnect(),
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java]: Can we add a javadoc to `RegisteredRpcConnection#tryReconnect` stating that it only returns false in the case of a concurrent modification?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java [hunk 2]
```diff
@@ -230,18 +231,20 @@ public void reconnect(final JobID jobId) {
 	 */
 	private final class JobManagerLeaderListener implements LeaderRetrievalListener {
```
Comment (TisonKun) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java]: `@ThreadSafe` since it is.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/JobLeaderServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+
+/**
+ * Tests for the {@link JobLeaderService}.
+ */
+public class JobLeaderServiceTest extends TestLogger {
+
+	@ClassRule
+	public static TestingRpcServiceResource rpcServiceResource = new TestingRpcServiceResource();
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/JobLeaderServiceTest.java]: `final`

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java [hunk 4]
```diff
@@ -230,18 +231,20 @@ public void reconnect(final JobID jobId) {
...
 
+		private final Object lock = new Object();
+
 		/** Job id identifying the job to look for a leader. */
 		private final JobID jobId;
 
 		/** Rpc connection to the job leader. */
-		private volatile RegisteredRpcConnection<JobMasterId, JobMasterGateway, JMTMRegistrationSuccess> rpcConnection;
+		private RegisteredRpcConnection<JobMasterId, JobMasterGateway, JMTMRegistrationSuccess> rpcConnection;
```
Comment (TisonKun) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java]: Possibly `@GuardedBy("lock")`

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java [hunk 5]
```diff
@@ -251,90 +254,83 @@ private JobManagerLeaderListener(JobID jobId) {
 		}
 
 		public void stop() {
-			stopped = true;
+			synchronized (lock) {
+				stopped = true;
```
Comment (TisonKun) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java]: break if `false`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java [hunk 6]
```diff
@@ -251,90 +254,83 @@ private JobManagerLeaderListener(JobID jobId) {
...
+					LOG.debug("New leader information for job {}. Address: {}, leader id: {}.",
+						jobId, leaderAddress, jobMasterId);
+
+					if (leaderAddress == null || leaderAddress.isEmpty()) {
+						// the leader lost leadership but there is no other leader yet.
+						if (rpcConnection != null) {
 							rpcConnection.close();
+						}
 
+						jobManagerLostLeadership = Optional.of(currentJobMasterId);
```
Comment (TisonKun) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java]: `ofNullable` seems required if you think about the first execution.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java [hunk 7]
```diff
@@ -251,90 +254,83 @@ private JobManagerLeaderListener(JobID jobId) {
...
-					if (stopped) {
-						rpcConnection.close();
-					} else {
 						LOG.info("Try to register at job manager {} with leader id {}.", leaderAddress, leaderId);
 						rpcConnection.start();
 					}
 				}
 			}
+
+			// send callbacks outside of the lock scope
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java]: why is this required/desirable?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java (hunk 6 above)
Comment (TisonKun) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java]: Will I get that the `ZooKeeperLeaderRetrievalService` helps filter switch from `null` to `null`, but it seems no harm we change it to `ofNullable` for encapsulating the implementation. Anytime we don't find a job manager lost leadership we just skip sending callbacks.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java [hunk 8]
```diff
@@ -432,8 +432,8 @@ protected void onRegistrationFailure(Throwable failure) {
 		@Override
 		protected CompletableFuture<RegistrationResponse> invokeRegistration(
 				JobMasterGateway gateway,
-				JobMasterId jobMasterId,
-				long timeoutMillis) throws Exception {
+				JobMasterId fencingToken,
```
Comment (TisonKun) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java]: Most where we will call it `jobMasterId` though. I'm ok with this change.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java (hunk 7 above)
Comment (TisonKun) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java]: It sends a message to TM actor. This action needs not to be synchronized.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java (hunk 2 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java]: Good point. Will add it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java (hunk 4 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java]: Sounds good.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java (hunk 5 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java]: Yes, but I think this wouldn't make much of a difference.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java (hunk 6 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java]: You are right. I will change it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java (hunk 8 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java]: Ok, then I keep as is because it is defined with this name in the super class.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java (hunk 1 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java]: Yes, good idea. I'll add it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/JobLeaderServiceTest.java (hunk 3 above)
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/JobLeaderServiceTest.java]: Good idea.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java (hunk 7 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/JobLeaderService.java]: Exactly. Triggering callbacks outside of the lock scope is preferrable because one does not know what the callback will do. It could happen that it waits for another lock which could then lead to a deadlock.

**Comment by tillrohrmann** [comment]
Thanks for the review @TisonKun and @zentol. I've addressed your comments and updated the PR accordingly.

**Comment by tillrohrmann** [comment]
Thanks for the review @zentol and @TisonKun. Merging this PR now.
